src/main/java/dev/inspector/agent/executor/Inspector.java

```diff
@@ -41,44 +59,55 @@ public Segment startSegment(String type, String label){
 
     public Segment startSegment(String type){
         if(!isRecording()){
-            throw new java.lang.Error("No active transaction found");
+            throw new IllegalStateException("No active transaction found");
         }
-        Segment segment = new Segment(transaction.getBasicTransactionInfo(), type);
+
+        Transaction currentMonitoringTransaction = transactionThreadSafeWrapper.get();
+        LOGGER.debug("Thread {}: Starting segment of type {} for transaction {}",
+                Thread.currentThread().getName(),
+                type,
+                currentMonitoringTransaction.getBasicTransactionInfo().getHash()
+        );
+        Segment segment = new Segment(currentMonitoringTransaction.getBasicTransactionInfo(), type);
+        segment.start();
         addEntries(segment);
         return segment;
     }
 
     public boolean isRecording(){
-        return transaction != null;
+        return transactionThreadSafeWrapper.get() != null;
     }
 
 
     public void addEntries(Transportable data) {
-        while(transport.getQueueSize() >= conf.getMaxEntries()){
-            transport.flush();
+        if (dataTransportService.getQueueSize() >= inspectorConfig.getMaxEntries()) {
+            dataTransportService.flush();
```

Comment:
I have done the changes you suggested. I think that with these changes you might encounter a situation in the future in which segments are not added in a transaction due to the transport queue being full.

if this would happen maybe the addEntries method might be modified to also include flushing like it did before